### PR TITLE
fix(django): handle composed queries in postgres

### LINKED ISSUE #2001: TypeError: can not serialize 'Composed' object
### Which version of dd-trace-py are you using?

0.46, having upgraded from 0.43

### Which version of the libraries are you using?

`protobuf==3.14.0`
`tenacity==6.3.1`

and relevantly:

`psycopg2-binary==2.8.6`

### How can we reproduce your problem?

Send a trace with a psycopg2

### What is the result that you get?

On Sentry after deploying the ddtrace upgrade, immediately got:

`TypeError: can not serialize 'Composed' object`

This looks like a recurrence of #732.

Stack frames:

ddtrace/internal/_encoding.pyx in ddtrace.internal._encoding.Packer._pack at line 263
ddtrace/internal/_encoding.pyx in ddtrace.internal._encoding.Packer._pack at line 187
ddtrace/internal/_encoding.pyx in ddtrace.internal._encoding.Packer.pack at line 269
ddtrace/internal/_encoding.pyx in ddtrace.internal._encoding.Packer.pack at line 272
ddtrace/internal/_encoding.pyx in ddtrace.internal._encoding.MsgpackEncoder.encode_trace at line 300
ddtrace/internal/_encoding.pyx in ddtrace.internal._encoding.MsgpackEncoder.encode_trace at line 299
ddtrace/internal/writer.py in write at line 257

### What is the result that you expected?

No error.

### LINKED ISSUE #2438: Proposal: Avoid nested db spans in django integration
# Background

The Django integration includes additional instrumentation for Django database backends. The Django database instrumentation results in two spans for any query executions if the database library has also been patched. For instance, if a mysql integration has been enabled in the tracer, a user will see the following in their traces:

image

The implementation of this is handled as follows:
1. the `django.db.connections` collection is [patched](https://github.com/DataDog/dd-trace-py/blob/d5b84311858870411c4a4c13486dc32b8e29477f/ddtrace/contrib/django/patch.py#L101) such that any new connection is [patched](https://github.com/DataDog/dd-trace-py/blob/d5b84311858870411c4a4c13486dc32b8e29477f/ddtrace/contrib/django/patch.py#L95) so that the cursor function will [wrapped](https://github.com/DataDog/dd-trace-py/blob/d5b84311858870411c4a4c13486dc32b8e29477f/ddtrace/contrib/django/patch.py#L85) with the `TracedCursor` provided by the `dbapi` integration.
2. the `TracedCursor` wraps a cursor object such that a [query execution will within a span](https://github.com/DataDog/dd-trace-py/blob/d5b84311858870411c4a4c13486dc32b8e29477f/ddtrace/contrib/dbapi/__init__.py#L76).
3. if the db cursor wrapped with `TracedCursor` by the Django integration is already wrapped, then there will be two spans created, the parent span created by the `TracedCursor` from the Django integration and the child span created by the wrapped cursor from any patched database library.

# Motivation

## Unnecessary

* The Django span timing is not meaningfully different.
* The Django span metadata is redundant.

## Bugs

* Coercion: while the psycopg integration [coerces](https://github.com/DataDog/dd-trace-py/blob/d5b84311858870411c4a4c13486dc32b8e29477f/ddtrace/contrib/psycopg/patch.py#L60) a composed query, the db span that is created by Django does not perform the same coercion when it's setting the span resource. The `TracedCursor` from the dbapi integration [sets the resource](https://github.com/DataDog/dd-trace-py/blob/d5b84311858870411c4a4c13486dc32b8e29477f/ddtrace/contrib/dbapi/__init__.py#L62) to the `query` without any coercion. If this coercion does not occur, the encoding of the trace using our custom msgpack encoder will cause an error to be raised. While the psycopg integration has been fixed to address this bug with the custom msgpack encoder, the same was not applied to the dbapi integration nor the Django db backend instrumentation.

# Proposal

1. Change the Django db instrumentation so as to avoid wrapping a cursor object that has already been wrapped.
2. This will mean that Django users by default will not see two spans for db calls.

### PR DESCRIPTION
**Pull request by majorgreys**
## Description

The Django db instrumentation wraps Django's own cursor wrapper with the dbapi `TracedCursor`. But this presents a problem for the case of postgres where the query, which is used to set the span resource, is not a string but instead an internal psycopg2 composable sql object. The psycopg2 integration provides its own cursor wrapper, which we can use in the Django db instrumentation if we determine that it applies. This resolves the problem with composable sql objects.

Backport to 0.49

Closes #2438 
Fixes #2001 
Blocked on #2499

## Checklist
- [ ] Added to the correct milestone.
- [X] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

**Comment by majorgreys** [comment]
@Mergifyio backport 0.49

File: ddtrace/contrib/django/patch.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/django/patch.py:93]: is vendor guaranteed to be `"postgresql"` for psycopg cursors? 

would we be able to do something like `isinstance(cursor, psycopg.cursor.Cursor)` to be more certain?

File: ddtrace/contrib/django/patch.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/django/patch.py:113]: 🤔 is this change supposed to be here?

File: ddtrace/contrib/django/patch.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/django/patch.py:92]: super nit (feel free to ignore): maybe makes more sense to structure it as a multiline if-statement since there could be other cursor cases in the future

File: tests/contrib/django/test_django_snapshots.py
Comment (Kyle-Verhoog) [inline tests/contrib/django/test_django_snapshots.py:75]: 👍

**Review by Kyle-Verhoog — COMMENTED** [review]
🙌 awesome! Just two questions but looks great!

File: ddtrace/contrib/django/patch.py
Comment (majorgreys) [reply ddtrace/contrib/django/patch.py:93]: The vendor attribute on the postgresql database wrapper has been `postgres` for quite some time, even with the older module that was deprecated in 2.0: https://github.com/django/django/blob/stable/1.8.x/django/db/backends/postgresql_psycopg2/base.py#L68

But, I do like your suggestion to check the cursor we get from the wrapped func call and wrap accordingly.

Comment (brettlangdon) [reply ddtrace/contrib/django/patch.py:93]: 👍🏻 agree this might be too brittle, especially if a different postgresql client were used

File: ddtrace/contrib/django/patch.py
Comment (brettlangdon) [reply ddtrace/contrib/django/patch.py:113]: old branch/bad merge from master, since we did remove this "recently" right?

**Review by brettlangdon — COMMENTED** [review]
agree with Kyle's feedback, otherwise lgtm

File: ddtrace/contrib/django/patch.py
Comment (majorgreys) [reply ddtrace/contrib/django/patch.py:113]: So the reason for this was mainly because of problems with testing multiple db connections. But if you try and start a setup the django app with multiple db backends, our patching does not get applied to all connections:

```
>>> import django
>>> django.setup()
>>> django.db.connections.__setitem__
<FunctionWrapper at 0x103febf40 for method at 0x103fb4580>
>>> django.db.connections["postgres"].cursor
<bound method BaseDatabaseWrapper.cursor of <django.db.backends.postgresql.base.DatabaseWrapper object at 0x103dfe4f0>>
>>> django.db.connections["default"].cursor
<FunctionWrapper at 0x104003ee0 for method at 0x103ff3ec0>
``` 

I can pull this change out into a separate PR

Comment (Kyle-Verhoog) [reply ddtrace/contrib/django/patch.py:113]: Oh interesting, yeah makes sense to pull into a separate PR IMO

**Review by Kyle-Verhoog — COMMENTED** [review]
LGTM once #2499 is merged

**Review by brettlangdon — DISMISSED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

File: ddtrace/contrib/django/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/django/patch.py]: do we need to protect `cursor.cursor`? what is `cursor` was `None` or something unexpected?

File: tests/snapshots/tests.contrib.django.test_django_snapshots.test_psycopg_query_default.snap
Comment (brettlangdon) [inline tests/snapshots/tests.contrib.django.test_django_snapshots.test_psycopg_query_default.snap:1]: are we going to have issues with using the same operation name?

is this changing at all from what we do today?

File: ddtrace/contrib/django/patch.py
Comment (Kyle-Verhoog) [reply ddtrace/contrib/django/patch.py]: this is a good point, might be worth 🦆-typing

File: tests/snapshots/tests.contrib.django.test_django_snapshots.test_psycopg_query_default.snap
Comment (majorgreys) [reply tests/snapshots/tests.contrib.django.test_django_snapshots.test_psycopg_query_default.snap:1]: this operation name is set within the django integration here:

https://github.com/DataDog/dd-trace-py/blob/c09c4dddab5cbc4222233a9ab0f286eaeac5261a/ddtrace/contrib/django/patch.py#L79https://github.com/DataDog/dd-trace-py/blob/c09c4dddab5cbc4222233a9ab0f286eaeac5261a/ddtrace/contrib/django/patch.py#L84
https://github.com/DataDog/dd-trace-py/blob/c09c4dddab5cbc4222233a9ab0f286eaeac5261a/ddtrace/contrib/dbapi/__init__.py#L40-L41

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**
